lib/room.cpp

```diff
@@ -2323,9 +2544,15 @@ void Room::uploadFile(const QString& id, const QUrl& localFilename,
                     d->fileTransfers[id].update(sent, total);
                     emit fileTransferProgress(id, sent, total);
                 });
-        connect(job, &BaseJob::success, this, [this, id, localFilename, job] {
+        connect(job, &BaseJob::success, this, [this, id, localFilename, job, encryptedFile] {
             d->fileTransfers[id].status = FileTransferInfo::Completed;
-            emit fileTransferCompleted(id, localFilename, QUrl(job->contentUri()));
+            if (encryptedFile) {
+                auto file = *encryptedFile;
+                file.url = QUrl(job->contentUri());
+                emit fileTransferCompleted(id, localFilename, QUrl(job->contentUri()), file);
+            } else {
+                emit fileTransferCompleted(id, localFilename, QUrl(job->contentUri()));
+            }
```

Comment:
...you could just explicitly pass it here, making the code more concise with avoided rewrapping of the `Omittable`.
```suggestion
            auto file = encryptedFile;
            if (file) {
                file->url = QUrl(job->contentUri());
            }
            emit fileTransferCompleted(id, localFilename, QUrl(job->contentUri()), file);
```
Didn't test but I think it should work?

By the way, passing templated types through signals is precarious when it comes to supporting QML - it is generally recommended to make an alias and use that instead. Also, `Omittable<>` is not known to the Qt type system for now. Not necessary to fix in this PR.

Reply:
Could probably even be
```suggestion
            if (encryptedFile) {
                encryptedFile->url = QUrl(job->contentUri());
            }
            emit fileTransferCompleted(id, localFilename, QUrl(job->contentUri()), encryptedFile);
```

Reply:
Argh, sorry, I forgot that I explicitly suppressed `Omittable::operator->() &`. I think I used to have `editValue()` for that but deleted it because nothing used it. For now it's a bit clumsy one-liner then:
```suggestion
            lift([&job] (EncryptedFile& ef) { ef.url = QUrl(job->contentUri(); }, encryptedFile);
            emit fileTransferCompleted(id, localFilename, QUrl(job->contentUri()), file);
```
I'll take care of that after merging.

Reply:
I tried applying it, but after two additional fixes it still doesn't compile...:

```
libquotient/lib/omittable.h:33:39: error: no matching function for call to ‘invoke(Quotient::Room::uploadFile(const QString&, const QUrl&, const QString&)::<lambda()>::<lambda(Quotient::EncryptedFile&)>, const Quotient::Omittable<Quotient::EncryptedFile>::value_type&)’
   33 |                ? Omittable(std::invoke(std::forward<FnT>(fn), *args...))
      |                            ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/qt5/QtCore/qhashfunctions.h:48,
                 from /usr/include/qt5/QtCore/QHashFunctions:1,
                 from /home/tobias/kde/src/libquotient/lib/util.h:10,

/usr/include/c++/12/type_traits:3034:11: error: no type named ‘type’ in ‘struct std::invoke_result<Quotient::Room::uploadFile(const QString&, const QUrl&, const QString&)::<lambda()>::<lambda(Quotient::EncryptedFile&)>, const Quotient::EncryptedFile&>’
 3034 |     using invoke_result_t = typename invoke_result<_Fn, _Args...>::type;
      |           ^~~~~~~~~~~~~~~
/home/tobias/kde/src/libquotient/lib/room.cpp: In function ‘Quotient::RoomEventPtr makeRedacted(const Quotient::RoomEvent&, const Quotient::RedactionEvent&)’:
/home/tobias/kde/src/libquotient/lib/room.cpp:2687:34: warning: ‘static Quotient::event_type_t Quotient::RoomMemberEvent::typeId()’ is deprecated: Use RoomMemberEvent::TypeId directly instead [-Wdeprecated-declarations]
 2687 |         { RoomMemberEvent::typeId(), { QStringLiteral("membership") } },
```